src/mlir/design.md

```diff
@@ -0,0 +1,158 @@
+
+```
+module @__ {
+  func @____FooBar__Foo__foo() {
+    ...
+  }
+}
+```
+
+And all calls to that function would also be changed in every other function that
+calls this one.
+
+## Load/Store Semantics
+
+Stack variables are allocated by LLVM IR with an `alloca` instruction. MLIR has
+a similar concept and we use that in the first implementation. The main problem
+with this approach is that there is only one dialect that has alloca/load/store:
+`memref`.
+
+MemRefs represent the memory used by `tensors`, which are multi-dimensional
+objects (matrices) of the same type. Verona is obviously not restricted to
+tensors as types, so we'll need to extend this to structure types, with complex
+layout semantics (ex. union types). For that, we'll need a set of dialect
+instructions that can be later lowered directly into LLVM for allocating stack
+space on complex layout semantics.
+
+For now, every function argument and local declaration (`let`, `var`) are
+allocated in the stack and its value stored in the memory space. This simplifies
+code generation (all variable symbols point to an address), but it doesn't help
+with value symbols and are really slow.
+
+The LLVM pipeline, even at O1, recognises most of the argument, temporary and
+local allocas as bogus and optimises them away, so this isn't a critical issue
+from a performance point of view, but we will probably need to update the symbol
+table soon to account for more types and states of the variables we hold (ex.
+`let` vs `var`).
+
+## Runtime Calls
+
+The Verona runtime is where the magic happens. Allocating regions, choosing the
+memory management strategy, scheduling behaviours, dispatching foreign language
+code, etc. The AST, however, has only the semantics of the language (with hints
+to the runtime behaviour), so we need to translate it at the MLIR lowering level.
+
+For example, heap allocation, which LLVM uses `malloc` calls, Verona uses
+`snmalloc`, which has a finer control over what we can do with the regions. In
+the end, however, it will be the same: calls to the runtime library in IR.
+
+The generator knows what runtime calls are appropriate for each AST construct,
+insert those calls in the right places and makes sure to declare the methods
+so that linking with the runtime library works.
+
+## Arithmetic
+
+Arithmetic is generated by the generator via pattern matching on the types and
+"method" names. Adding two numbers (`a + b`) is the same as calling the method
+`+` on `a` over `b` (`a.+(b)`), but these methods don't exist (how would we
+implement addition if not with a `+`?).
+
+So the generator has to find those that weren't declared and replace with
+arithmetic if the types are compatible (int, float).
+
+Users are still allowed to use `+` as a method name, in which case the generator
+will recognise and call that method instead of transforming into arithmetic, even
+if the types are compatible. For example:
```

Comment:
Please can you describe what happens when you cast an integer to an interface type that defines a `+` method?  We'll need to have a `+` method that can be called via the vtable from somewhere, is this in the standard library?  In which case, we end up with code that looks something like this:

```verona
class U32
{
  static +(a: U32, B: U32) { a + b; }
}
```

Where the method appears to be infinitely recursive, but the compiler will generate the special version for the version in the body.  Or does the compiler synthesise these methods on demand?
}

Reply:
All of that will have already been resolved by the parser and the AST will be unambiguous. What happens in the parser and if that's even legal is a matter for the parser to discuss, not the MLIR lowering of the AST.

Reply:
I don't think I understand how this can be implemented with the lowering that you've described, so it would be good to state this in the document.

Reply:
Sorry, I misunderstood your question. You are right, we can't implement that in any reasonable (non recursive) way. Or at least I can't see a way that that would work as "expected".

However, this isn't any different than any infinitely recursive function that you can create, if you change the name from `+` to anything else. So, in a way, the programmer gets what they want, given the syntax and semantics of the language.

What could work is something like:
```
class U32
{
  static +(a: U32, b: U32) { @llvm.math.funky_add(a, b); }
}
```
which is meaningful but very low level to allow programmers to do that, so it would be probably done by the standard included modules.

A way to make sure users don't break that, for example, would be to mandate the inclusion of the standard modules for every project, so you don't need to add "using numeric", to your programs. This way, programmers wouldn't be able to create their own `U32` classes, because it would be a re-definition of an existing one, working around the problem entirely.

I'm not saying this is a good idea, I'm just saying this is a high-level problem that is best fixed above the MLIR generator level.